Make more Regex loops atomic automatically

stephentoub opened the pull request:
For example, given an expression like "(abcd*)*e", the inner d* loop will now be converted to be atomic, e.g. "(abc(?>d*))*e)".  We're careful not to convert an expression like "(abca*)*e)", where the inner loop has no overlap with what comes immediately after the loop but does with the start of the loop.

I also fixed what appears to be a long-standing bug in how character classes are canonicalized.  An attempt was made to avoid running the canonlicalization routine, but we almost always need to do it, and we were not doing it in some cases, resulting in character classes that were correct but not as optimized as they should have been, e.g. two abutting ranges when one would suffice, leading to poorer code gen.  It's part of this PR because I caught when changing the reduction tests to validate the full code tree including strings rather than just the codes list.

Fixes https://github.com/dotnet/runtime/issues/27402
cc: @danmosemsft, @eerhardt, @ViktorHofer

stephentoub committed:
Change more Regex loops inside of atomic nodes to be atomic

stephentoub committed:
Allow Regex loops that end other loops to be made atomic

For example, given an expression like "(abcd*)*e", the inner d* loop will now be converted to be atomic, e.g. "(abc(?>d*))*e)".  We're careful not to convert an expression like "(abca*)*e)", where the inner loop has no overlap with what comes immediately after the loop but does with the start of the loop.

I also fixed what appears to be a long-standing bug in how character classes are canonicalized.  An attempt was made to avoid running the canonlicalization routine, but we almost always need to do it, and we were not doing it in some cases, resulting in character classes that were correct but not as optimized as they should have been, e.g. two abutting ranges when one would suffice, leading to poorer code gen.  It's part of this PR because I caught when changing the reduction tests to validate the full code tree including strings rather than just the codes list.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:
```diff
@@ -997,6 +997,17 @@ public void IsMatch_Invalid()
 
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)] // take too long due to backtracking
+        [Theory]
+        [InlineData(@"(\w*)+\.", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", false)]
```
I think this is what I was thinking of when I opened the issue. 

Does this help a `(\w*)+` match succeed -- I assume that still runs interminably?

stephentoub replied:
> Does this help a (\w*)+ match succeed

Yes. This doesn't eliminate all backtracking, but it changes `(\w*)+` to `((?>\w*))+`, which means that it won't backtrack into the `\w+`.  It'll still do a lot of backtracking, but also a lot less than it otherwise would.

Note that it doesn't help with the original expression off of which you based this, the one commented out in the tests.  That one has catastrophic backtracking, and there's little we can do about that; try it out on regex101 and you'll see the same terribleness with pcre.

danmosemsft commented:
https://github.com/dotnet/docs/pull/16980

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
+        {
+            if (maxDepth <= 0)
+            {
+                // We hit our recursion limit.  Just don't apply the optimization.
```
I assume we have tests that hit this

stephentoub replied:
I'll double-check and add tests to ensure we do.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
-                                    return false;
-                                }
-                            }
+            // If the loop's body is a concatenate, we can skip to its last child iff that
```
For comments like this I'm mentally visualizing patterns, you might consider sprinkling more example patterns in comments.

stephentoub replied:
I've gone through and added a lot of such comments.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
+            {
+                int concatCount = node.ChildCount();
+                RegexNode lastConcatChild = node.Child(concatCount - 1);
+                if (CanBeMadeAtomic(lastConcatChild, node.Child(0), maxDepth - 1))
```
In this case, we're passing in the last child of the "concat" as the node, and the first child of the "concat" in as the subsequent. Maybe subsequent is not the right term then -- referent?

stephentoub replied:
The idea is that we're testing it as if it's the subsequent node, e.g. with ab* testing what happens if it were b*a.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
-                                        return true;
-                                }
-                                break;
+            // Skip the successor down to the guaranteed next node.
```
Not changed line, but it's not clear what is meant by "guaranteed next node". Maybe "closest node that's guaranteed to follow it"  ?

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -312,6 +280,77 @@ internal RegexNode FinalOptimize()
+                    case Concatenate:
+                        // For Alternate, Loop, and Lazyloop, make the node atomic by wrapping it in an Atomic node.
+                        // Since we later check to see whether a node is atomic based on its parent or grandparent,
+                        // we don't bother wrapping such a node in an Atomic one.
```
```suggestion
                        // we don't bother wrapping such a node in an Atomic one if its parent is already Atomic.
```

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -312,6 +280,77 @@ internal RegexNode FinalOptimize()
+                        continue;
+
+                    // For Loop, we search to see if there's a viable last expression, and iff there
+                    // is we recur into processing it.
```
```suggestion
                    // is we recurse into processing it.
```

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -312,6 +280,77 @@ internal RegexNode FinalOptimize()
+                            RegexNode? loopDescendent = FindLastExpressionInLoopForAutoAtomic(node, maxDepth - 1);
+                            if (loopDescendent != null)
+                            {
+                                node = loopDescendent;
```
I assumed "descendent" was a spelling error, but apparently it's not. 😛

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1155,26 +1191,41 @@ private void ReduceConcatenateWithAutoAtomic()
+                    while (true)
                     {
-                        node = node.Child(node.ChildCount() - 1);
+                        // We can always recur into captures and into the last node of concatenations.
```
```suggestion
                        // We can always recurse into captures and into the last node of concatenations.
```

danmosemsft replied:
Well, maybe "recur" is a possible word to use, but "recurse" is more common. Either is fine I guess.

stephentoub replied:
I had it drilled into me in college that "recur" is what happens when you employ recursion, and "recourse" is what happens when you swear again. Ever since "recurse" just sounds wrong to me.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
+            // Skip past captures.
+            while (node.Type == Capture)
+            {
+                node = node.Child(0);
```
this is guaranteed to never lead to a null `node` ?

stephentoub replied:
Yes, a capture will always have a child.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/tests/Regex.Groups.Tests.cs:
```diff
@@ -730,6 +730,8 @@ public static IEnumerable<object[]> Groups_Basic_TestData()
             yield return new object[] { null, @"(?:a{2}?){3}?", "aaaaaaaaa", RegexOptions.None, new string[] { "aaaaaa" } };
             yield return new object[] { null, @"(?:(?:[ab]c[de]f){3}){2}", "acdfbcdfacefbcefbcefbcdfacdef", RegexOptions.None, new string[] { "acdfbcdfacefbcefbcefbcdf" } };
             yield return new object[] { null, @"(?:(?:[ab]c[de]f){3}hello){2}", "aaaaaacdfbcdfacefhellobcefbcefbcdfhellooooo", RegexOptions.None, new string[] { "acdfbcdfacefhellobcefbcefbcdfhello" } };
+            // Nested atomic
```
A nested atomic to me means one atomic group inside another ? Like your `[InlineData("(?>(?>(?>(?:abc)*)))", "(?:abc)*")]`

stephentoub replied:
I meant things nested inside an atomic that could/should themselves be made atomic.

stephentoub commented:
Heh, CI failed because this ComponentModel test is expecting catastrophic backtracking that it's not getting anymore :smile:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/RegularExpressionAttributeTests.cs#L78-L83

stephentoub committed:
Address PR feedback

And add more comments, more code coverage, more runtime asserts, etc.

I also noticed we weren't reducing Setloopatomic and Setlazy into {Not}oneloopatomic and [Not}onelazy, so I fixed that.

And I changed ReduceAtomic to actually remove unnecessary intermediate atomic nodes.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -184,15 +184,108 @@ private void MakeLoopAtomic()
-                    break;
                 default:
-                    Debug.Fail($"Unexpected type: {Type}");
+#if DEBUG
```
Why is `#if DEBUG` required here? My understanding is that in release, `Debug.Assert` and everything that calls it is elided, and this includes any string interopolation being passed to it. Is that not the case?

stephentoub replied:
The TypeName property is under an `#if DEBUG`.  In a release build, it's as if here this was `Debug.Assert(condition. $"{SomePropertyThatDoesntActuallyExist}")`, which doesn't compile.

danmosemsft replied:
ah yes

danmosemsft commented:
Your example patterns in comments are super helpful.
